mmgen/models/architectures/sngan_proj/modules.py

```diff
@@ -0,0 +1,236 @@
+from copy import deepcopy
+
+import numpy as np
+import torch.nn as nn
+from mmcv.cnn import (ConvModule, build_activation_layer, build_norm_layer,
+                      build_upsample_layer, constant_init, xavier_init)
+
+from mmgen.models.builder import MODULES
+from mmgen.utils import check_dist_init
+
+
+@MODULES.register_module()
+class SNGAN_GenResBlock(nn.Module):
+
+    # _default_conv_cfg = dict(kernel=3, stride=1, padding=1, act_cfg=None)
+    # _default_shortcut_cfg = dict(kernel=1, stride=1, padding=0, act_cfg=None)
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 hidden_channels=None,
+                 num_classes=0,
+                 use_cbn=True,
+                 use_norm_para=False,
+                 act_cfg=dict(type='ReLU'),
+                 norm_cfg=dict(type='BN'),
+                 upsample_cfg=dict(type='nearest'),
+                 upsample=True,
+                 auto_sync_bn=True,
+                 conv_cfg=None,
+                 shortcut_cfg=None):
+
+        super().__init__()
+        self.learnable_sc = in_channels != out_channels or upsample
+
+        self.activateion = build_activation_layer(act_cfg)
+        hidden_channels = out_channels if hidden_channels is None \
+            else hidden_channels
+
+        if upsample:
+            self.upsample = build_upsample_layer(upsample_cfg)
+
+        conv_blocks = [
+            ConvModule(in_channels, hidden_channels, **conv_cfg),
+            ConvModule(hidden_channels, out_channels, **conv_cfg)
+        ]
+        self.conv_blocks = nn.ModuleList(conv_blocks)
+
+        norm_blocks = [
+            SNcBatchNorm(in_channels, num_classes, use_cbn, norm_cfg,
+                         use_norm_para, auto_sync_bn),
+            SNcBatchNorm(hidden_channels, num_classes, use_cbn, norm_cfg,
+                         use_norm_para, auto_sync_bn)
+        ]
+        self.norm_blocks = nn.ModuleList(norm_blocks)
+
+        if self.learnable_sc:
+            self.shortcut = ConvModule(in_channels, out_channels,
```

Comment:
I may suggest removing the shortcut config because this setting may be fixed in this model. In addition, `with_spectral_norm` should be consistent with the `conv_blocks`.